### PR TITLE
Use Absolute-indexes as keys for the Env-mapping during closure-conversion

### PR DESCRIPTION
**Pull request by nickchapman-da**
Done
- Use Absolute-indexes as keys for the `Env.mapping` during closure-conversion.
- [Do runtime check to confirm behaviour matches the existing Relative-indexes, on full repo. See 1st commit]
- remove Relative-index keys for `Env.mapping`
- remove quadratic `shiftLoc`
- improve `ClosureConversion` stack-safety tests

This addresses aspect (2) of the quadratic complexity of `shiftLoc`, described in issue #11830.

#### Benchmarking

shows a small, 3%, performance improvement in speedy compilation with this change (Actually, multiply that percentage by 3, because the speedy compilation benchmark is dominated by type checking which takes twice as long as the compilation -- fixed in coming PR: #11927).

base:
```
  SpeedyCompilation.bench             avgt  200  0.197 ±  0.001   s/op
```
change:
```
  SpeedyCompilation.bench             avgt  200  0.191 ± 0.001   s/op
```

### CODE REVIEW

**Review by cocreature — APPROVED** [review]
Nice, thank you!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversion.scala [hunk 1]
```diff
@@ -20,42 +20,45 @@ private[speedy] object ClosureConversion {
...
-        // We just update the keys of the map (the relative-indexes from the original SEVar)
-        val m1 = mapping.map { case (k, loc) => (n + k, loc) }
-        // And create mappings for the `n` new stack items
-        val m2 = (1 to n).view.map { rel =>
-          val abs = this.depth + n - rel
-          (rel, target.SELocAbsoluteS(abs))
+      def extend(n: Int): Env = {
+        // Create mappings for `n` new stack items, and combine with the (unshifted!) existing mapping.
+        val m2 = (1 to n).view.map { i =>
+          val abs = Abs(this.sourceDepth + (n - i))
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversion.scala:38]:
```suggestion
          val abs = Abs(this.sourceDepth + n - i)
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversion.scala [hunk 2]
```diff
@@ -20,42 +20,45 @@ private[speedy] object ClosureConversion {
...
+          (abs, target.SELocAbsoluteS(this.targetDepth + n - i))
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversion.scala:39]: curious to know why we subtract `i` from `n`, and not use something like :
```suggestion
        val m2 = (0 until n).view.map { i =>
          val abs = Abs(this.sourceDepth + i)
          (abs, target.SELocAbsoluteS(this.targetDepth + i))
```

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversion.scala (hunk 2 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversion.scala:39]: Yeah. This was a hang-over from the old _relative_ indexing which started from `1`.
The new _absolute_ indexing starts from `0`

I simplified the code along the lines you suggest in both the `extend` and `absBody` methods.
Much simpler now. Thanks!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversion.scala [hunk 3]
```diff
@@ -20,42 +20,45 @@ private[speedy] object ClosureConversion {
...
-      def shift(n: Int): Env = {
-        // We just update the keys of the map (the relative-indexes from the original SEVar)
-        val m1 = mapping.map { case (k, loc) => (n + k, loc) }
-        // And create mappings for the `n` new stack items
-        val m2 = (1 to n).view.map { rel =>
-          val abs = this.depth + n - rel
-          (rel, target.SELocAbsoluteS(abs))
+      def extend(n: Int): Env = {
+        // Create mappings for `n` new stack items, and combine with the (unshifted!) existing mapping.
+        val m2 = (0 to n - 1).view.map { i =>
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversion.scala:37]:
```suggestion
        val m2 = (0 until n).view.map { i =>
```

Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ClosureConversion.scala]: ok!

**Pull request merged**